dotnet-desktop-guide/framework/winforms/controls/how-to-iterate-through-all-nodes-of-a-windows-forms-treeview-control.md

```diff
@@ -13,111 +14,34 @@ ms.assetid: 427f8928-ebcf-4beb-887f-695b905d5134
-       foreach (TreeNode n in nodes)  
-       {  
-          PrintRecursive(n);  
-       }  
-    }  
-    ```  
-  
-    ```cpp  
-    private:  
-       void PrintRecursive( TreeNode^ treeNode )  
-       {  
-          // Print the node.  
-          System::Diagnostics::Debug::WriteLine( treeNode->Text );  
-          MessageBox::Show( treeNode->Text );  
-  
-          // Print each node recursively.  
-          System::Collections::IEnumerator^ myNodes = (safe_cast<System::Collections::IEnumerable^>(treeNode->Nodes))->GetEnumerator();  
-          try  
-          {  
-             while ( myNodes->MoveNext() )  
-             {  
-                TreeNode^ tn = safe_cast<TreeNode^>(myNodes->Current);  
-                PrintRecursive( tn );  
-             }  
-          }  
-          finally  
-          {  
-             IDisposable^ disposable = dynamic_cast<System::IDisposable^>(myNodes);  
-             if ( disposable != nullptr )  
-                      disposable->Dispose();  
-          }  
-       }  
-  
-       // Call the procedure using the TreeView.  
-       void CallRecursive( TreeView^ treeView )  
-       {  
-          // Print each node recursively.  
-          TreeNodeCollection^ nodes = treeView->Nodes;  
-          System::Collections::IEnumerator^ myNodes = (safe_cast<System::Collections::IEnumerable^>(nodes))->GetEnumerator();  
-          try  
-          {  
-             while ( myNodes->MoveNext() )  
-             {  
-                TreeNode^ n = safe_cast<TreeNode^>(myNodes->Current);  
-                PrintRecursive( n );  
-             }  
-          }  
-          finally  
-          {  
-             IDisposable^ disposable = dynamic_cast<System::IDisposable^>(myNodes);  
-             if ( disposable != nullptr )  
-                      disposable->Dispose();  
-          }  
-       }  
-    ```  
+:::code language="csharp" source="snippets/how-to-iterate-through-all-nodes-of-a-windows-forms-treeview-control/cs/Form1.cs" id="PrintRecursive":::
+
+:::code language="vb" source="snippets/how-to-iterate-through-all-nodes-of-a-windows-forms-treeview-control/vb/Form1.vb" id="PrintRecursive":::  
+
+:::code language="cpp" source="snippets/how-to-iterate-through-all-nodes-of-a-windows-forms-treeview-control/cpp/MyForm.h" id="PrintRecursive":::  
   
+## Non-recursive approach
+
+Recursive code may lead to stack overflow errors or out of memory exceptions due to there being too many frames on the call stack.
+
+The following example is an alternate iterative approach to traversing the nodes of the tree using a <xref:System.Collections.Generic.Queue%601> data structure. This approach does not follow any ordering and only ensures all the nodes are printed. If we wish to follow a pre-order or post-order approach then we can use a <xref:System.Collections.Generic.Stack%601> data structure to ensure that a node and its children are processed before subsequent nodes.
```

Comment:
Moved the warning into the recursive section. Cleaned this up a bit. I think it's a bit more clear now, thoughts?

```suggestion
The following example is an alternate iterative approach to traversing the nodes of the tree using a <xref:System.Collections.Generic.Queue%601> collection. This approach doesn't follow the parent-child relationship of a node only ensures every node is printed. If you want to process each tree node and its children, first, use the <xref:System.Collections.Generic.Stack%601> collection
```